### PR TITLE
GUAC-246: Implement screen recording

### PR DESCRIPTION
**Pull request by mike-jumper**

### CODE REVIEW

File: src/guacenc/guacenc.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    /* Load defaults */
+    int width = GUACENC_DEFAULT_WIDTH;
+    int height = GUACENC_DEFAULT_HEIGHT;
+    int bitrate = GUACENC_DEFAULT_BITRATE;
+
+    /* Parse arguments */
+    int opt;
+    while ((opt = getopt(argc, argv, "s:r:")) != -1) {
+
+        /* -d: Dimensions */
```
Comment (mike-jumper) [inline src/guacenc/guacenc.c]: Oops. This should be `-s`.

File: src/guacenc/guacenc.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * the command line. This name is dictated by ffmpeg / libavcodec.
+ */
+#define GUACENC_DEFAULT_CODEC "mpeg4"
+
+/**
+ * The extension to append to the end of the input file to produce the output
+ * file name, excluding the separating period, if no other suffix is specified
+ * on the command line.
+ */
+#define GUACENC_DEFAULT_SUFFIX "mpg"
```
Comment (mike-jumper) [inline src/guacenc/guacenc.h]: Neither `GUACENC_DEFAULT_CODEC` nor `GUACENC_DEFAULT_SUFFIX` are used - they are from an earlier iteration where both of these were configurable.

File: src/guacenc/image-stream.h [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+#define GUACENC_IMAGE_STREAM_H
+
+#include "config.h"
+#include "buffer.h"
+
+#include <cairo/cairo.h>
+
+/**
+ * The initial number of bytes to allocate for the image data buffer. If this
+ * buffer is not sufficiently larged, it will be dynamically reallocated as it
```
Comment (mike-jumper) [inline src/guacenc/image-stream.h]: "larged" ... :confounded:

File: src/guacenc/display-buffers.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+}
+
+int guacenc_display_free_buffer(guacenc_display* display,
+        int index) {
+
+    /* Transform index to buffer space */
+    int internal_index = -index - 1;
+
+    /* Do not lookup / allocate if index is invalid */
```
Comment (jmuehlner) [inline src/guacenc/display-buffers.c]: I don't see any allocation done here?

File: src/guacenc/display-flatten.c [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        if (cairo == NULL)
+            continue;
+
+        /* Render buffer to layer */
+        cairo_reset_clip(cairo);
+        cairo_rectangle(cairo, layer->x, layer->y, src->width, src->height);
+        cairo_clip(cairo);
+
+        cairo_set_source_surface(cairo, surface, layer->x, layer->y);
+        cairo_paint_with_alpha(cairo, layer->opacity / 255.0);
```
Comment (jmuehlner) [inline src/guacenc/display-flatten.c]: Maybe this 255.0 should be a little bit less magical?

File: src/guacenc/display-buffers.c (hunk 4 above)
Comment (mike-jumper) [inline src/guacenc/display-buffers.c]: Trueeeee...

File: src/guacenc/display-layers.c [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    /* Current layer depth is the depth of the parent + 1 */
+    return guacenc_display_get_depth(display, parent) + 1;
+
+}
+
+int guacenc_display_free_layer(guacenc_display* display,
+        int index) {
+
+    /* Do not lookup / allocate if index is invalid */
```
Comment (jmuehlner) [inline src/guacenc/display-layers.c]: Also not allocating here..

File: src/guacenc/display-flatten.c (hunk 5 above)
Comment (mike-jumper) [inline src/guacenc/display-flatten.c]: I'm not sure I agree. All color components in Guacamole, including alpha, are documented as having a maximum of `0xFF` (255) and a minimum of `0x00`, and that maximum is a common convention within graphics. In Cairo, these range between `0.0` and `1.0`. We're simply converting the ranges here.

File: src/guacenc/image-stream.h [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+ * of the Guacamole video encoder display; the allocated guacenc_image_stream
+ * will not automatically be associated with the active display, nor will the
+ * provided layer/buffer index be validated.
+ *
+ * @param mask
+ *     The Guacamole protocol compositing operation (channel mask) to apply
+ *     when drawing the image.
+ *
+ * @param index
+ *     The index of the layer or bugger that the image should be drawn to.
```
Comment (jmuehlner) [inline src/guacenc/image-stream.h]: Bugger?

File: src/guacenc/display.h [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+ * @param index
+ *     The index of the stream to create. All valid stream indices are
+ *     non-negative.
+ *
+ * @param mask
+ *     The Guacamole protocol compositing operation (channel mask) to apply
+ *     when drawing the image.
+ *
+ * @param layer_index
+ *     The index of the layer or bugger that the image should be drawn to.
```
Comment (jmuehlner) [inline src/guacenc/display.h]: bugger......

File: src/guacenc/display-flatten.c (hunk 5 above)
Comment (jmuehlner) [inline src/guacenc/display-flatten.c]: Ok, sounds good to me.

**Pull request merged**